File: Python/textbook_track/chapter09/ba9a.py
# Construct a Trie from a Collection of Patterns
import itertools as it
import sys

from collections import defaultdict


class Trie:
    def __init__(self, patterns):
        self._node_id_generator = it.count()
        self._root = next(self._node_id_generator)
        self._adjacency_list = self._build_trie_from_patterns(patterns)

    def edges(self):
        for node, neighbours in self._adjacency_list.items():
            for neighbour, label in neighbours:
                yield f'{node}->{neighbour}:{label}'

    def _build_trie_from_patterns(self, patterns):
        adjacency_list = defaultdict(list)
        for pattern in patterns:
            self._add_pattern_to_trie(adjacency_list, pattern)
        return dict(adjacency_list)

    def _add_pattern_to_trie(self, adjacency_list, pattern):
        current_node = self._root
        for letter in pattern:
            node = _get_neighbour_with_given_label(adjacency_list, current_node, letter)
            if node is None:
                next_node = next(self._node_id_generator)
                adjacency_list[current_node].append((next_node, letter))
                current_node = next_node
            else:
                current_node = node


def _get_neighbour_with_given_label(adjacency_list, current_node, letter):
    neighbours = adjacency_list.get(current_node, [])
    for node, label in neighbours:
        if label == letter:
            return node
    return None


def main():
    patterns = sys.stdin.read().splitlines()
    trie = Trie(patterns)
    for edge in trie.edges():
        print(edge)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9A.scala
package TextbookTrack.Chapter09

object BA9A {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}

  private type Node = Int

  final case class Edge(nodeFrom: Node, nodeTo: Node, label: Char) {
    override def toString: String = s"$nodeFrom->$nodeTo:$label"
  }

  class Trie(patterns: List[String]) {
    import Trie.getNeighbourWithGivenLabel

    private val nodeIdGenerator: Iterator[Int] = Iterator.from(0)
    private val root: Int = nodeIdGenerator.next()
    private val adjacencyList: Map[Node, List[Edge]] = buildTrieFromPatterns()

    def edges(): Iterator[Edge] = adjacencyList.valuesIterator.flatten

    private def buildTrieFromPatterns(): Map[Node, List[Edge]] = {
      val adjacencyList: MutableMap[Node, List[Edge]] = MutableMap()
      patterns.foreach(addPatternToTrie(adjacencyList, _))
      adjacencyList.toMap
    }

    private def addPatternToTrie(adjacencyList: MutableMap[Node, List[Edge]], pattern: String): Unit = {
      @tailrec
      def loop(letters: List[Char], currentNode: Node): Unit = letters match {
        case Nil => ()
        case letter :: lss =>
          getNeighbourWithGivenLabel(adjacencyList, currentNode, letter) match {
            case None =>
              val nextNode: Node = nodeIdGenerator.next()
              val neighbours: List[Edge] = adjacencyList.getOrElse(currentNode, Nil)
              adjacencyList.update(currentNode, Edge(currentNode, nextNode, letter) :: neighbours)
              loop(lss, nextNode)
            case Some(node) => loop(lss, node)
        }
      }

      loop(pattern.toList, root)
    }
  }

  object Trie {
    private def getNeighbourWithGivenLabel(adjacencyList: MutableMap[Node, List[Edge]],
                                           currentNode: Node,
                                           letter: Char): Option[Node] = {
      adjacencyList
        .get(currentNode)
        .flatMap(_.find{ case Edge(_, _, label) => label == letter })
        .map(_.nodeTo)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val patterns: List[String] = reader.toList
    val trie = new Trie(patterns)
    trie.edges().foreach(println)
  }
}
